Dev

rkpamegan opened the pull request:
Resolves  https://github.com/cadop/dhart/issues/73
- `get_node_attributes` has optional `ids` argument which is passed to native `c_get_node_attributes` and processed
- Gives attributes for the specified IDs if provided, otherwise gives attributes for all IDs
- Cinterface function calls one of two `GetNodeAttributes` functions based on whether or not `ids` is provided
- Added example code for C++ `GetNodeAttributes`, Python `get_node_attributes`
- Short unittests for Cinterface and C++ functions
- Updated description on all affected functions

cadop commented on src/CMakeSettings.json:
```diff
@@ -1,180 +1,223 @@
+            ]
+        },
+        {
+            "name": "x64-Release-CSharp",
```
looks like this config is the same as the previous one.   Can you rebase the change or undo a whitespace change?

cadop commented on src/Cpp/tests/src/SpatialStructures.cpp:
```diff
@@ -1927,6 +1936,16 @@ namespace GraphExampleTests {
 		}
+
+		vector<int> ids = { 0, 3 };
+		auto subset_attrs = g.GetNodeAttributes(ids, "cross slope");
```
can you use a different name than `cross slope` for this demo attr.

rkpamegan replied:
will be renamed to "testattribute"

cadop commented on src/Csharp/CMakeLists.txt:
```diff
@@ -69,7 +69,7 @@ set_target_properties(
 		PROPERTIES
 		VS_DOTNET_REFERENCES "${normal_references}"
-		VS_DOTNET_TARGET_FRAMEWORK_VERSION "v4.7.2"
+		VS_DOTNET_TARGET_FRAMEWORK_VERSION "v4.5"
```
Why was this decremented?

rkpamegan replied:
was not intentional, will be reverted

cadop commented on src/Cpp/tests/src/SpatialStructures.cpp:
```diff
@@ -2066,10 +2085,26 @@ namespace CInterfaceTests {
 
+		std::vector<int> subset_ids = { 3, 5, 7 };
+
+		char** scores_out2 = new char* [subset_ids.size()];
```
can you use a better name than `scores_out2` for this section?

rkpamegan replied:
will be renamed to `subset_scores_out`, along with renaming for size variable

cadop commented on src/Cpp/tests/src/SpatialStructures.cpp:
```diff
@@ -1024,8 +1024,17 @@ namespace EdgeExampleTests {
 		HF::SpatialStructures::Edge edge(node, score, STEP::NOT_CONNECTED);
 	}
+
+	TEST(_edge, EdgeConstructorThing) {
```
Fix test name.

rkpamegan replied:
I don't think this was meant to be an actual change, and it seems to be the same as the test directly above, so I'll just delete.

cadop commented on src/Cpp/tests/src/SpatialStructures.cpp:
```diff
@@ -2028,7 +2047,7 @@ namespace CInterfaceTests {
 		// By the postconditions of GetNodeAttributes, this should update scores_out,
 		// and scores_out_size with the variables we need
-		GetNodeAttributes(&g, attr_type.c_str(), scores_out, &scores_out_size);
+		GetNodeAttributes(&g, NULL, attr_type.c_str(), 0, scores_out, &scores_out_size);
```
Split this test into a new one, so the old one stays the same, but a new one allows this additional argument.

cadop commented on src/Cinterface/spatialstructures_C.cpp:
```diff
@@ -319,16 +319,30 @@ C_INTERFACE AddNodeAttributes(
 }
 
-C_INTERFACE GetNodeAttributes(const HF::SpatialStructures::Graph* g, const char* attribute, 
+C_INTERFACE GetNodeAttributes(const HF::SpatialStructures::Graph* g, const int* ids, const char* attribute, int num_nodes,
```
Is this a breaking change to the c interace?

rkpamegan replied:
I changed every usage of this function (in `spatial_structures_native_functions.c_get_node_attributes` and the C++/Cinterface tests) to fit the updated parameters.

cadop reviewed (changes requested):
Could you also post a screenshot of the C# and C++ unittest framework passing in visual studio.

rkpamegan commented:
![image](https://github.com/user-attachments/assets/26adaea1-0326-42c8-bfcb-fe9feb3a8f89)

Reverted `GetNodeAttributes <_graphCInterface>` and `GetNodeAttributes <_graph>`, added tests for the new functions `GetNodeAttributesByID <_graphCInterface>` and `GetNodeAttributesByID <_graph>`
